lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1462 @@
+ * @bin: a #EknMediaBin
+ *
+ * Return the media URI
+ */
+EMB_DEFINE_GETTER (const gchar *, uri, NULL)
+
+/**
+ * ekn_media_bin_set_uri:
+ * @bin: a #EknMediaBin
+ * @uri:
+ *
+ * Sets the media URI to play
+ */
+EMB_DEFINE_SETTER_STRING (uri, URI,
+  /* Make playbin show the first video frame if there is an URI
+   * and the widget is realized.
+   */
+   ekn_media_bin_set_state_ready (bin);
+)
+
+/**
+ * ekn_media_bin_get_autohide:
+ * @bin: a #EknMediaBin
+ *
+ * Returns control's auto hide timeout in seconds.
+ */
+EMB_DEFINE_GETTER (gint, autohide, 0)
+
+/**
+ * ekn_media_bin_set_autohide:
+ * @bin: a #EknMediaBin
+ * @autohide: A timeout in seconds
+ *
+ * Sets the timeout to auto hide controls
+ */
+EMB_DEFINE_SETTER (gint, autohide, AUTOHIDE,)
+
+/**
+ * ekn_media_bin_get_fullscreen:
+ * @bin: a #EknMediaBin
+ *
+ * Returns whether video is fullscreen or not
+ */
+EMB_DEFINE_GETTER (gboolean, fullscreen, FALSE)
+
+/**
+ * ekn_media_bin_set_fullscreen:
+ * @bin: a #EknMediaBin
+ * @fullscreen:
+ *
+ * Sets whether to show the video in fullscreen mode or not
+ */
+EMB_DEFINE_SETTER (gboolean, fullscreen, FULLSCREEN,
+  /* If there is no video sink, delay fullscreen until realize event */
+  if (bin->video_sink)
+    ekn_media_bin_fullscreen_apply (bin, fullscreen);
+)
+
+/**
+ * ekn_media_bin_get_title:
+ * @bin: a #EknMediaBin
+ *
+ * Returns the media title if any
+ */
+EMB_DEFINE_GETTER (const gchar *, title, NULL)
+
+/**
+ * ekn_media_bin_set_title:
+ * @bin: a #EknMediaBin
+ * @title:
+ *
+ * Sets the media title.
+ * By default EknMediaBin will use the title from the media metadata
+ */
+EMB_DEFINE_SETTER_STRING (title, TITLE,
+  gtk_label_set_label (bin->top_left_label, title);
+)
+
+/**
+ * ekn_media_bin_get_description:
+ * @bin: a #EknMediaBin
+ *
+ * Returns the media description if any
+ */
+EMB_DEFINE_GETTER (const gchar *, description, NULL)
+
+/**
+ * ekn_media_bin_set_description:
+ * @bin: a #EknMediaBin
+ * @description:
+ *
+ * Sets the media description.
+ * By default EknMediaBin will use the description from the media metadata
+ */
+EMB_DEFINE_SETTER_STRING (description, DESCRIPTION,)
+
+/**
+ * ekn_media_bin_get_volume:
+ * @bin: a #EknMediaBin
+ *
+ */
+gdouble
+ekn_media_bin_get_volume (EknMediaBin *bin)
+{
+  g_return_val_if_fail (EKN_IS_MEDIA_BIN (bin), 1.0);
+  return gtk_scale_button_get_value (bin->volume_button);
+}
+
+/**
+ * ekn_media_bin_set_volume:
+ * @bin: a #EknMediaBin
+ * @volume: from 0.0 to 1.0
+ *
+ */
+void
+ekn_media_bin_set_volume (EknMediaBin *bin, gdouble volume)
+{
+  g_return_if_fail (EKN_IS_MEDIA_BIN (bin));
+  gtk_scale_button_set_value (bin->volume_button, volume);
+  g_object_notify_by_pspec (G_OBJECT (bin), properties[PROP_VOLUME]);
+}
+
+/**
+ * ekn_media_bin_play:
+ * @bin: a #EknMediaBin
+ *
+ */
+void
+ekn_media_bin_play (EknMediaBin *bin)
+{
+  g_return_if_fail (EKN_IS_MEDIA_BIN (bin));
+
+  g_object_set (bin->play, "uri", bin->uri, NULL);
+
+  bin->state = GST_STATE_PLAYING;
```

Comment:
Its not for performance actually, ->state is meant to keep the user desired state, which is not always the same as the pipeline state.
I am probably going ahead of me here since we do not really need it right now, but I am pretty sure we will eventually.

Say we are streaming some video and the buffer runs out so we have to pause the pipeline, then once the buffer catch up, we wont know which stated the used wanted :S

In any case it deserves a comment :)